engine/servermaster/jobmanager.go

```diff
@@ -289,17 +264,124 @@ func (jm *JobManagerImplV2) SubmitJob(ctx context.Context, req *pb.SubmitJobRequ
+	}
+	if req.Job.Id != "" && !jobNameRegex.MatchString(req.Job.Id) {
+		return status.Errorf(codes.InvalidArgument, "job id must match %s", jobNameRegex.String())
+	}
+	if req.Job.Type == pb.Job_TypeUnknown {
+		return status.Error(codes.InvalidArgument, "job type must be specified")
+	}
+	return nil
+}
+
+// ListJobs implements JobManagerServer.ListJobs.
+func (jm *JobManagerImpl) ListJobs(ctx context.Context, req *pb.ListJobsRequest) (*pb.ListJobsResponse, error) {
+	masterMetas, err := jm.frameMetaClient.QueryJobs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Slice(masterMetas, func(i, j int) bool {
+		return masterMetas[i].ID < masterMetas[j].ID
+	})
+
+	resp := &pb.ListJobsResponse{
+		NextPageToken: req.PageToken,
+	}
+	firstIdx := sort.Search(len(masterMetas), func(i int) bool {
+		return masterMetas[i].ID > req.PageToken
+	})
+	for i := firstIdx; i < len(masterMetas); i++ {
+		if masterMetas[i].Tp == framework.JobManager {
+			continue
+		}
+		job, err := buildPBJob(masterMetas[i])
+		if err != nil {
+			return nil, err
+		}
+		resp.NextPageToken = job.Id
+		resp.Jobs = append(resp.Jobs, job)
+		if req.PageSize > 0 && int32(len(resp.Jobs)) >= req.PageSize {
+			break
+		}
+	}
+
+	return resp, nil
+}
+
+func buildPBJob(masterMeta *frameModel.MasterMetaKVData) (*pb.Job, error) {
+	var jobType pb.Job_Type
+	switch tp := framework.MustConvertWorkerType2JobType(masterMeta.Tp); tp {
+	case engineModel.JobTypeCVSDemo:
+		jobType = pb.Job_CVSDemo
+	case engineModel.JobTypeDM:
+		jobType = pb.Job_DM
+	case engineModel.JobTypeCDC:
+		jobType = pb.Job_CDC
+	case engineModel.JobTypeFakeJob:
+		jobType = pb.Job_FakeJob
+	default:
+		return nil, errors.Errorf("job %s has unknown type %v", masterMeta.ID, masterMeta.Tp)
+	}
+
+	var jobStatus pb.Job_Status
+	switch masterMeta.StatusCode {
+	case frameModel.MasterStatusUninit:
+		jobStatus = pb.Job_Created
+	case frameModel.MasterStatusInit:
+		jobStatus = pb.Job_Running
+	case frameModel.MasterStatusFinished:
+		jobStatus = pb.Job_Finished
+	case frameModel.MasterStatusStopped:
+		jobStatus = pb.Job_Canceled
+	default:
+		return nil, errors.Errorf("job %s has unknown type %v", masterMeta.ID, masterMeta.StatusCode)
+	}
+
+	return &pb.Job{
+		Id:     masterMeta.ID,
+		Type:   jobType,
+		Status: jobStatus,
+		Config: masterMeta.Config,
+		Error:  nil, // TODO: Fill error field.
+	}, nil
+}
+
+// GetJobMasterInfo implements JobManager.GetJobMasterInfo.
+func (jm *JobManagerImpl) GetJobMasterInfo(ctx context.Context, jobID string) (*JobMasterInfo, error) {
+	// FIXME: The JobFsm's status may not be consistent with the database.
+	job := jm.JobFsm.QueryOnlineJob(jobID)
+	if job == nil {
+		if _, err := jm.frameMetaClient.GetJobByID(ctx, jobID); pkgOrm.IsNotFoundError(err) {
+			return nil, ErrJobNotFound.GenWithStack(&JobNotFoundError{JobID: jobID})
+		}
+		return nil, ErrJobNotRunning.GenWithStack(&JobNotRunningError{JobID: jobID})
+	}
```

Comment:
We have stored executor address in `master_meta_kv_data`, can we use it directly.
- Then we don't need to query executor address from `executorManager`
- and can also rename this function to a more explicit name, such as `GetJobMasterForwardAddress`

Reply:
Done.